generative/inferers/inferer.py

```diff
@@ -571,6 +569,639 @@ def get_likelihood(
+            controlnet: Callable[..., torch.Tensor],
+            cn_cond: torch.Tensor,
+            scheduler: Callable[..., torch.Tensor] | None = None,
+            save_intermediates: bool | None = False,
+            intermediate_steps: int | None = 100,
+            conditioning: torch.Tensor | None = None,
+            mode: str = "crossattn",
+            verbose: bool = True,
+            seg: torch.Tensor | None = None,
+    ) -> torch.Tensor | tuple[torch.Tensor, list[torch.Tensor]]:
+        """
+        Args:
+            input_noise: random noise, of the same shape as the desired sample.
+            diffusion_model: model to sample from.
+            controlnet: controlnet sub-network.
+            cn_cond: conditioning image for the ControlNet.
+            scheduler: diffusion scheduler. If none provided will use the class attribute scheduler
+            save_intermediates: whether to return intermediates along the sampling change
+            intermediate_steps: if save_intermediates is True, saves every n steps
+            conditioning: Conditioning for network input.
+            mode: Conditioning mode for the network.
+            verbose: if true, prints the progression bar of the sampling process.
+            seg: if diffusion model is instance of SPADEDiffusionModel, segmentation must be provided.
+        """
+        if mode not in ["crossattn", "concat"]:
+            raise NotImplementedError(f"{mode} condition is not supported")
+
+        if not scheduler:
+            scheduler = self.scheduler
+        image = input_noise
+        if verbose and has_tqdm:
+            progress_bar = tqdm(scheduler.timesteps)
+        else:
+            progress_bar = iter(scheduler.timesteps)
+        intermediates = []
+        for t in progress_bar:
+            # 1. ControlNet forward
+            down_block_res_samples, mid_block_res_sample = controlnet(
+                x=image, timesteps=torch.Tensor((t,)).to(input_noise.device),
+                controlnet_cond= cn_cond,
+            )
+            # 2. predict noise model_output
+            if mode == "concat":
+                model_input = torch.cat([image, conditioning], dim=1)
+                if isinstance(diffusion_model, SPADEDiffusionModelUNet):
+                    model_output = diffusion_model(
+                        model_input, timesteps=torch.Tensor((t,)).to(input_noise.device), context=None, seg=seg,
+                        down_block_additional_residuals=down_block_res_samples,
+                        mid_block_additional_residual=mid_block_res_sample
+                    )
+                else:
+                    model_output = diffusion_model(
+                        model_input, timesteps=torch.Tensor((t,)).to(input_noise.device), context=None,
+                        down_block_additional_residuals=down_block_res_samples,
+                        mid_block_additional_residual=mid_block_res_sample
+                    )
```

Comment:
```suggestion
                _seg = seg if isinstance(diffusion_model, SPADEDiffusionModelUNet) else None
                model_output = diffusion_model(
                    model_input, timesteps=torch.Tensor((t,)).to(input_noise.device), context=None, seg=_seg,
                    down_block_additional_residuals=down_block_res_samples,
                    mid_block_additional_residual=mid_block_res_sample
                )
```

We could reduce the amount of code duplication with this sort of pattern, here and elsewhere.

Reply:
that exact formulation won't work as `DiffusionModelUnet` doesn't have the argument `seg`. But this would work:

```python
    diffusion_model = partial(diffusion_model, seg=seg) if isinstance(diffusion_model, SPADEDiffusionModelUNet) else diffusion_model
    model_output = diffusion_model(
        model_input, timesteps=torch.Tensor((t,)).to(input_noise.device), context=None,
        down_block_additional_residuals=down_block_res_samples,
        mid_block_additional_residual=mid_block_res_sample

```

(would need `from functools import partial` up top)